lib/codecs/src/encoding/format/csv.rs

```diff
@@ -0,0 +1,184 @@
+    pub const fn new(csv: CsvSerializerOptions) -> Self {
+        Self { csv }
+    }
+
+    /// Build the `CsvSerializer` from this configuration.
+    pub fn build(&self) -> Result<CsvSerializer, BuildError> {
+        if self.csv.fields.is_empty() {
+            Err("At least one CSV field must be specified".into())
+        } else {
+            Ok(CsvSerializer::new(self.csv.fields.clone()))
+        }
+    }
+
+    /// The data type of events that are accepted by `CsvSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // CSV, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerOptions {
+    /// The CSV Serializer fields.
+    ///
+    /// The order of the fields determines the order of the columns in the CSV output.
+    /// Make sure to use the same order when parsing the CSV output.
+    ///
+    /// If a field is not present in the event, the output will be `NaN`.
+    ///
+    /// Some value types: Array, Regex, Object are not supported by the CSV format.
+    /// If a field is of an unsupported type, the output will be `NaN`.
+    pub fields: Vec<ConfigOwnedValuePath>,
+}
+
+/// Serializer that converts an `Event` to bytes using the CSV format.
+#[derive(Debug, Clone)]
+pub struct CsvSerializer {
+    fields: Vec<ConfigOwnedValuePath>,
+}
+
+impl CsvSerializer {
+    /// Creates a new `CsvSerializer`.
+    pub const fn new(fields: Vec<ConfigOwnedValuePath>) -> Self {
+        Self { fields }
+    }
+}
+
+impl Encoder<Event> for CsvSerializer {
+    type Error = vector_common::Error;
+
+    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        let log = event.into_log();
+        let values = log.value();
+        let mut wtr = csv::Writer::from_writer(vec![]);
+
+        for field in &self.fields {
+            match values.get(field) {
+                Some(Value::Bytes(bytes)) => wtr.write_field(String::from_utf8(bytes.to_vec())?)?,
+                Some(Value::Integer(int)) => wtr.write_field(int.to_string())?,
+                Some(Value::Float(float)) => wtr.write_field(float.to_string())?,
+                Some(Value::Boolean(bool)) => wtr.write_field(bool.to_string())?,
+                Some(Value::Timestamp(timestamp)) => wtr.write_field(timestamp.to_rfc3339())?,
+                Some(Value::Null) => wtr.write_field("NaN")?,
+                // Other value types: Array, Regex, Object are not supported by the CSV format.
+                Some(_) => wtr.write_field("NaN")?,
+                None => wtr.write_field("NaN")?,
```

Comment:
Do we really want to write `NaN` if the field doesn't exist at all? I've been told that this could conceivably be OK, but I know I've also seen non-existent fields simply be skipped over (i.e. you'd write an empty string, so the output would look like `.. some_other_value,,another_value ..`) and skipping the field _feels_ like the right thing to do if it doesn't exist at all.

Reply:
That seems more reasonable, updated.